File: Olympiad/COCI/official/2008/contest3_solutions/cudak.pas
program cudak;

var
   A, B : int64;
   pow10 : array [0..15] of int64;
   memo : array [0..15,0..135] of int64;
   minSolution : int64;
   i, j, S : longint;

function rec( prefix : int64; digits, sum : longint ) : int64;
var ret, mini, maxi : int64;
    dig : longint;
begin
   mini := prefix;
   maxi := prefix + pow10[digits]-1;

   if (sum < 0) or (mini > B) or (maxi < A) then begin
      rec := 0;
      exit;
   end;

   if digits = 0 then begin
      if sum > 0 then
         rec := 0
      else begin
         rec := 1;
         if minSolution = -1 then minSolution := prefix;
      end;

      exit;
   end;

   if (mini >= A) and (maxi <= B) and (memo[digits,sum] <> -1) then begin
      rec := memo[digits,sum];
      exit;
   end;

   ret := 0;
   for dig := 0 to 9 do
      ret := ret + rec( prefix + dig*pow10[digits-1], digits-1, sum-dig );

   if (mini >= A) and (maxi <= B) then memo[digits,sum] := ret;
   rec := ret;
end;

begin
   pow10[0] := 1;
   for i := 1 to 15 do pow10[i] := pow10[i-1] * 10;

   readln( A, B, S );

   for i := 0 to 15 do
      for j := 0 to 135 do
         memo[i,j] := -1;
   minSolution := -1;

   writeln( rec( 0, 15, S ) );
   writeln( minSolution );
end.

File: Olympiad/COCI/official/2008/contest3_solutions/cudak.cpp
#include <cstdio>
#include <cstring>

typedef long long llint;

llint A, B;
llint pow10[16];
llint memo[16][136];
llint min_solution = -1;

llint rec( llint prefix, int digits, int sum ) {
   if( sum < 0 ) return 0;

   llint mini = prefix;
   llint maxi = prefix + pow10[digits]-1;
   if( mini > B || maxi < A ) return 0;

   if( digits == 0 ) {
      if( sum > 0 ) return 0;      
      if( min_solution == -1 ) min_solution = prefix;
      return 1;
   }

   bool memoize = (mini >= A && maxi <= B);
   if( memoize && memo[digits][sum] != -1 ) return memo[digits][sum];

   llint ret = 0;
   for( int dig = 0; dig < 10; ++dig ) 
      ret += rec( prefix + dig*pow10[digits-1], digits-1, sum-dig );   

   if( memoize ) memo[digits][sum] = ret;
   return ret;
}

int main( void ) {
   pow10[0] = 1;
   for( int i = 1; i <= 15; ++i ) pow10[i] = pow10[i-1] * 10;
   
   int S;
   scanf( "%lld%lld%d", &A, &B, &S );
   
   memset( memo, -1, sizeof memo );

   printf( "%lld\n", rec( 0, 15, S ) );
   printf( "%lld\n", min_solution );

   return 0;
}
